### PR TITLE
[DRAFT] Add co-printing support

### PR DESCRIPTION
**Pull request by navruzm**

### CODE REVIEW

File: cmd/pack3d/main.go [hunk 1]
```diff
@@ -203,42 +290,94 @@ func main() {
...
 	/*
 		model.Mesh().SaveSTL(fmt.Sprintf("pack3d-%s.stl", os.Args[5]))
 		model.TreeMesh().SaveSTL(fmt.Sprintf("out%dtree.stl", int(score*100000)))
 	*/
 	done()
-}
\ No newline at end of file
+}
+
+type Config struct {
```
Comment (far-authentise) [inline cmd/pack3d/main.go:354]: 👍

File: cmd/pack3d/main.go [hunk 2]
```diff
@@ -203,42 +290,94 @@ func main() {
...
 	}
 	fmt.Println("the fill percentage is:", fillPercentage)
-	ioutil.WriteFile(fmt.Sprintf("%s.json", os.Args[5]), positions_json, 0644)
-	//os.Stdout.Write(positions_json)
+	ioutil.WriteFile(fmt.Sprintf("%s.json", *fileNameArg), positions_json, 0644)
+	// os.Stdout.Write(positions_json)
+	// model.Mesh().SaveSTL(fmt.Sprintf("pack3d-%s.stl", *fileNameArg))
 
-	//TODO: Unblock the following line if want to generate the packing STL
+	// STL file is no longer created
```
Comment (far-authentise) [inline cmd/pack3d/main.go]:
```suggestion
	// STL file is no longer created, results returned as JSON for separate packer.
```

File: cmd/pack3d/main.go [hunk 3]
```diff
@@ -203,42 +290,94 @@ func main() {
...
-			transMatrix = [4][4]float64{{0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}}
+			transMaps = append(transMaps, TransMap{srcStlNames[j], transMatrix, 0})
+
+			m = meshMap[coprint.Filename]
+			m.Transform(fauxgl.Translate(success_model.Items[j].Translation))
+			t = pack3d.Rotations[success_model.Items[j].Rotation].Translate(success_model.Items[j].Translation)
+			transMatrix = [4][4]float64{{t.X00, t.X01, t.X02, t.X03}, {t.X10, t.X11, t.X12, t.X13}, {t.X20, t.X21, t.X22, t.X23}, {t.X30, t.X31, t.X32, t.X33}}
+			transMaps = append(transMaps, TransMap{coprint.Filename, transMatrix, 0})
+
+			fmt.Println("")
```
Comment (far-authentise) [inline cmd/pack3d/main.go]:
```suggestion
			# will be removed or commented out in  final release.
			fmt.Println("")
```

**Review by far-authentise — COMMENTED** [review]
Good start. Can you move the example we had into /tests directory? 

Do you know go unit tests standards? Please  make either a simple unit-test, or a manual command line test, given the example we sent you.

File: README.md [hunk 4]
```diff
@@ -10,7 +10,7 @@ Pack3d is written in golang and the installation instructions can be found in th
 ### Example
 
 ```
-pack3d {100,100,100} 5 output 1 mesh1.stl 1 mesh2.stl
+pack3d --json_file=input.json --build_volume=100,100,100
```
Comment (iras-auth) [inline README.md]: The `build_volume` would be better off inside of the  json file.

File: cmd/pack3d/main.go [hunk 5]
```diff
@@ -1,33 +1,34 @@
 /*
 Instruction: write down this into the command line to use the software
-<pack3d {your_frame_size_X,your_frame_size_Y,your_frame_size_Z} spacing your_output_name STL_numbers STL_path STL_numbers STL_path .....>
-For example: <pack3d {100,100,100} 5 Pikacu 1 /home/corner.stl 2 /home/Pika.stl>
+<pack3d --build_volume=your_frame_size_X,your_frame_size_Y,your_frame_size_Z --json_file=json_config_file --spacing=spacing --filename=export_filename>
```
Comment (iras-auth) [inline cmd/pack3d/main.go]: `spacing` and `build_volume` would be better placed in the json file.

File: tests/coprint/input.json [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+            "count": 1,
+            "coprint": {
+                "filename": "./tests/coprint/Merry.stl",
+                "transformation": [
+                    [1, 0, 0, -60],
+                    [0, 1, 0, -343],
+                    [0, 0, 1, 0],
+                    [0, 0, 0, 1]
+                ]
+            }
```
Comment (iras-auth) [inline tests/coprint/input.json]:
```suggestion
            "co_print": [
                {
                    "filename": "./tests/coprint/Merry.stl",
                    "transformation": [
                        [1, 0, 0, -60],
                        [0, 1, 0, -343],
                        [0, 0, 1, 0],
                        [0, 0, 0, 1]
                    ]
                }
            ]
```

Comment (iras-auth) [reply tests/coprint/input.json]: More than one object can be co-printed. The `Merry.stl` is just the first and only one in the list in this example.

File: cmd/pack3d/main.go [hunk 7]
```diff
@@ -1,33 +1,34 @@
...
+For example: <pack3d --json_file=input.json --build_volume=100,100,100>
 The unit of frame and spacing is millimeters.
 */
 
 package main
 
 import (
 	"encoding/json"
+	"flag"
 	"fmt"
```
Comment (iras-auth) [inline cmd/pack3d/main.go:13]: `	"gonum"`

File: cmd/pack3d/main.go [hunk 8]
```diff
@@ -1,33 +1,34 @@
...
 	"io/ioutil"
```
Comment (iras-auth) [inline cmd/pack3d/main.go:14]:
```suggestion
      "gonum"
	"io/ioutil"
```

Comment (iras-auth) [reply cmd/pack3d/main.go:14]: This library is necessary to do the inverse of a matrix, multiply two matrices and compute the determinant of a matrix.

https://pkg.go.dev/gonum.org/v1/gonum/mat#section-documentation

https://medium.com/wireless-registry-engineering/gonum-tutorial-linear-algebra-in-go-21ef136fc2d7

File: cmd/pack3d/main.go (hunk 5 above)
Comment (navruzm) [reply cmd/pack3d/main.go]: done!

File: cmd/pack3d/main.go (hunk 8 above)
Comment (navruzm) [reply cmd/pack3d/main.go:14]: It will automatically get imported if we use this library in the file.

File: README.md (hunk 4 above)
Comment (iras-auth) [inline README.md]: necessary to update this line in docs with just:

`pack3d --json_file=input.json`

File: cmd/pack3d/main.go [hunk 9]
```diff
@@ -40,80 +39,134 @@ func timed(name string) func() {
...
+					X20: cp.Transformation[2][0],
+					X21: cp.Transformation[2][1],
+					X22: cp.Transformation[2][2],
+					X23: cp.Transformation[2][3],
+					X30: cp.Transformation[3][0],
+					X31: cp.Transformation[3][1],
+					X32: cp.Transformation[3][2],
+					X33: cp.Transformation[3][3],
+				})
+				mesh.Add(coMesh)
```
Comment (iras-auth) [inline cmd/pack3d/main.go:152]: is this done twice for the same mesh, including line 173?

Comment (iras-auth) [reply cmd/pack3d/main.go:152]: No, my bad, please ignore the comment above, please.

File: cmd/pack3d/main.go [hunk 10]
```diff
@@ -40,80 +39,134 @@ func timed(name string) func() {
...
+			for _, cp := range item.Coprint {
+				coMesh, err := fauxgl.LoadMesh(cp.Filename)
+				if err != nil {
+					panic(err)
+				}
 
-		done = timed("centering mesh")
-		mesh.Center()
-		done()
+				coMesh.Transform(fauxgl.Matrix{
```
Comment (iras-auth) [inline cmd/pack3d/main.go:134]: This transform should be passed into `model.Add(mesh, bvhDetail, count, spacing, transform)` as an extra parameter.

Comment (iras-auth) [inline cmd/pack3d/main.go:134]: This transform, the co_packing_id of the object a mesh is co-packed with and a boolean should be passed into:

`model.Add(
     mesh,
     bvhDetail,
     count,
     spacing,
     is_co_packing,
     co_packing_transform,
     co_packing_id
)`

as extra parameters. We don't want to lose the original transform a mesh came with if that mesh had one.

Comment (iras-auth) [inline cmd/pack3d/main.go:134]:
```suggestion
				coMesh.parent_id = ...
				coMesh.Transform(fauxgl.Matrix{
```

Comment (iras-auth) [inline cmd/pack3d/main.go:133]:
```suggestion
				coMesh.parent_id = ...
				coMesh.co_packing_transform(fauxgl.Matrix{
```

File: cmd/pack3d/main.go [hunk 11]
```diff
@@ -152,13 +205,13 @@ func main() {
 		/* ntime is the times of trial to find a output solution, if after trying for 100 times
 		and no solution is found, then reset the model and try again. Usually if there is a solution,
 		ntime will be 1 or 2 for most cases. */
```
Comment (iras-auth) [inline cmd/pack3d/main.go:212]:
```suggestion
		ntime will be 1 or 2 for most cases. */
		for _, item := range config.Items {
		     if item.Coprint != nil {
		         /* step 1: find out the parent's transformation (given by the model.Pack(...)), by using the parent_id or whatever else.
		         /* step 2: override the item.Transform (given by the packing algorithm) with:
		         /*              a matrix multiplication of the parent's transform (found in step 1) by the item.co_packing_transform. The order of the multiplication is meaningful and will be dealt with later.
		     }
		}
```

File: tests/coprint/input.json [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+                    [0, 1, 0, -343],
+                    [0, 0, 1, 0],
+                    [0, 0, 0, 1]
+                ]
+            }]
+        },
+        {
+            "filename": "./tests/coprint/logo.stl",
+            "count": 1
+        }
```
Comment (iras-auth) [inline tests/coprint/input.json:38]:
```suggestion
        },
        {
             "filename": "./tests/coprint/Christmas.stl",
             "count": 2,
             "coprint": [{
                 "filename": "./tests/coprint/Merry.stl",
                 "transformation": [
                     [1, 0, 0, -500],
                     [0, 1, 0, -343],
                     [0, 0, 1, 0],
                     [0, 0, 0, 1]
                 ]
             }]
         },
```

File: tests/coprint/input.json [hunk 13]
```diff
@@ -0,0 +1,35 @@
+{
+    "build_volume": [1, 0, 0],
+    "spacing": 2,
+    "items": [{
+            "filename": "./tests/coprint/Christmas.stl",
```
Comment (iras-auth) [inline tests/coprint/input.json:6]:
```suggestion
            "filename": "./tests/coprint/Christmas_1.stl",
```

File: cmd/pack3d/main.go (hunk 8 above)
Comment (iras-auth) [reply cmd/pack3d/main.go:14]: Yes but as the code is meant to run in Docker containers, this dependency's name (and version number) need to be explicitly added in.

**Pull request closed without merging**